DOC Update Controllers section (#206)
silverstripe/developer-docs: en/02_Developer_Guides/02_Controllers/01_Introduction.md

# Introduction to Controllers

The following example is for a simple [`Controller`](api:SilverStripe\Control\Controller) class. When building off the Silverstripe CMS you will
subclass the base `Controller` class.

[info]

> If you're using the `cms` module with and dealing with [`SiteTree`](api:SilverStripe\CMS\Model\SiteTree) records then for your custom page controllers you

Reviewer: ```suggestion
If you're using the `cms` module and dealing with [`SiteTree`](api:SilverStripe\CMS\Model\SiteTree) records then for your custom page controllers you
```

Reviewer 2: Done

When choosing names for actions, avoid using the same name you've used for relations on the model the controller represents. If you have relations with the same name as controller actions, templates rendered for that controller which refer to the relation won't render as expected - they will attempt to render the action where you expect to be using the relation.

For example if the controller above was for a `Team` model which had a `Players` relation, the action should not also be named `players`. Something like `showPlayers` would be more appropriate.
[/warning]

## Routing

We need to define the URL that this controller can be accessed on. In our case, the `TeamsController` should be visible 
at http://yoursite.com/teams/ and the `players` custom action is at http://yoursite.com/team/players/.

[info]
If you're extending `ContentController` or `PageController` for your `SiteTree` records you don't need to define the routes value as the `cms` handles
routing for those.
[/info]

**app/_config/routes.yml**

```yml
---
Name: approutes
After: '#coreroutes'
---
SilverStripe\Control\Director:
  rules:
    'teams//$Action/$ID/$Name': 'App\Controller\TeamController'
```

[alert]

> Make sure that after you have modified the `routes.yml` file, that you clear your Silverstripe CMS caches using `?flush=1`.

Reviewer: ```suggestion
Make sure that after modifying the `routes.yml` file you clear your Silverstripe CMS caches using `?flush=1`.
```

Reviewer 2: Done

[/alert]

For more information about creating custom routes, see the [Routing](routing) documentation.

For information about how to get a URL (e.g. to use in templates) for a given controller action, see [Getting the URL for a controller action](#link) below.

## Actions

> Actions in controllers are specific routes which are accessible via HTTP. They are often backed by a method, though a method does not need to be declared so long as there

Reviewer: ```suggestion
Actions in controllers are specific routes which are accessible via HTTP or CLI requests. They are often backed by a method, though a method does not need to be declared so long as there
```

Reviewer 2: Done

is a template available to render for the given action.

> Controllers respond by default to an `index` action. You don't need to implement this action as a method method nor declare it as in `allowed_actions`, but you

Reviewer: ```suggestion
Controllers respond by default to an `index` action. You don't need to implement this action as a method nor declare it as in `allowed_actions`, but you
```

Reviewer 2: Done

can implement the `index()` method and return custom data to be used in the appropriate template (see [Template and Views](../templates)).

Action methods can return one of four things:

1. an array. The appropriate template will be rendered, using the values in the array you provided. The rendered result will be set as the body for the current [`HTTPResponse`](api:SilverStripe\Control\HTTPResponse) object.
2. a string (e.g. JSON or HTML markup). The string will be set as the body for the current `HTTPResponse` object.
3. an `HTTPResponse`. This can either be a new response or `$this->getResponse()`.
4. `$this` or `$this->customise()`. This will render the controller using the appropriate template and set the rendered result as the body for the current `HTTPResponse`.

[hint]

- returning `$this` is the equivalent of returning an empty array.
- returning `$this->customise()` is the equivalent of returning an array with data.

[/hint]

See [templates](#templates) below for information about declaring what template to use in the above scenarios.

A controller action can also throw an [`HTTPResponse_Exception`](api:SilverStripe\Control\HTTPResponse_Exception).

> This is a special exception that is indicates that a specific error HTTP code should be used in the response.

Reviewer: ```suggestion
This is a special exception that indicates that a specific error HTTP code should be used in the response.
```

Reviewer 2: Done

 */
public function ajaxaction() 
{
    $this->getResponse()->addHeader('Content-type', 'application/json');

    return json_encode([
        'json' => true,
    ]);
}
```

For more information on how a URL gets mapped to an action see the [Routing](routing) documentation.

## Security

See the [Access Control](access_control) documentation.

## Templates

The template to use for a given action is determined in the following order:

1. If a template has been explicitly declared for the action in the `templates` property, it will be used.
2. If a template has been explicitly declared for the "index" action in the `templates` property, it will be used (regardless of what action is being rendered).
3. If the `template` property has been set at all, its value will be used.
4. If a template exists with the name of this class or any of its ancestors, suffixed with the name of the action name, it will be used.

>     - e.g. for the `App\Control\TeamController` example, the "player" action would look for `templates/App/Control/TeamController_player.ss` and `templates/SilverStripe/Control/Controller_player.ss`.

Reviewer: to follow the recommended action name vs relation name
```suggestion
    - e.g. for the `App\Control\TeamController` example, the "showPlayers" action would look for `templates/App/Control/TeamController_showPlayers.ss` and `templates/SilverStripe/Control/Controller_showPlayers.ss`.
```

Reviewer 2: Done

    - Note that the "index" action skips this step.
5. If a template exists with the name of this class or any of its ancestors (with no suffix), it will be used.
    - e.g. for the `App\Control\TeamController` example, it would look for `templates/App/Control/TeamController.ss` and `templates/SilverStripe/Control/Controller.ss`.

[note]
Subclasses of `ContentController` additionally check for templates named similarly to the model the controller represents - for example a `HomePageController` class which represents a `HomePage` model will look for `HomePage_{action}.ss` after checking `HomePageController_{action}.ss`.
[/note]

You can declare templates to be used for an action by setting the `templates` array. The key should be the name of the action,
and the value should be a template name, or array of template names in cascading precedence.

```php
class TeamController extends Controller 
{
    protected $templates = [

>         'players' => 'TemplateForPlayers',

Reviewer: ```suggestion
        'showPlayers' => 'TemplateForPlayers',
```

Reviewer 2: Done

    ];

    private static $allowed_actions = [

>         'players',

Reviewer: ```suggestion
        'showPlayers',
```

Reviewer 2: Done

    ];
}
```

[warning]
The `templates` property is _not_ a configuration property, so if you declare it directly as in the above example you will
override any templates declared in parent classes. If you want to keep template declarations from parent classes, you could
apply new templates in a constructor like so:

```php
class TeamController extends SomeParentController 
{
    public function __construct()
    {
        parent::__construct();

>         $this->templates['players'] => 'TemplateForPlayers';

Reviewer: ```suggestion
        $this->templates['showPlayers'] => 'TemplateForPlayers';
```

Reviewer 2: Done

    }
}
```

[/warning]

> As mentioned in [Actions](#actions) above, controller actions return a string or `HTTPResponse` to bypass this template selection process.

Reviewer 2: ```suggestion
As mentioned in [Actions](#actions) above, controller actions can return a string or `HTTPResponse` to bypass this template selection process.
```

Reviewer 2: Done

For more information about templates, inheritance and how to render into views, See the 
[Templates and Views](../templates) documentation.

## Getting the URL for a controller action {#link}

Each controller should declare the `url_segment` configuration property, using the non-variable portion of that controller's routing rule.

```php
class TeamController extends Controller 
{
    private static $url_segment = 'teams';
}
```

You can then use the [`Link()`](api:SilverStripe\Control\RequestHandler::Link()) method to get a relative URL for your controller:

```php
$indexLink = $teamController::Link();
$playersActionLink = $teamController::Link('players');
```

You can of course also use `$Link` in a template.

[notice]
If you have more complex logic for determining the link for your controller, you can override the `Link()` method - in that case you should
be sure to invoke the `updateLink` extension method so that extensions can make changes as necessary: `$this->extend('updateLink', $link, $action);`
[/notice]

## Connecting Pages to ContentControllers

By default, a `SiteTree` subclass will be automatically associated with a controller which is in the same

> namespace as it, and is named the same but suffixed with the `Controller`. For example, `App\Page\HomePage`

Reviewer: ```suggestion
namespace, and is named the same but suffixed with the `Controller`. For example, `App\Page\HomePage`
```

Reviewer 2: Done

> namespace as it, and is named the same but suffixed with the `Controller`. For example, `App\Page\HomePage`

Reviewer 2: ```suggestion
namespace as it, and is named the same but suffixed with `Controller`. For example, `App\Page\HomePage`
```

Reviewer 2: Done

will be associated with a `App\Page\HomePageController` if such a class exists.

If there is no controller for a specific page class, that page's ancestors will be checked until a suitable

> controller is foun.d

Reviewer: ```suggestion
controller is found.
```

Reviewer 2: Done
